pkg/cloud/services/compute/instance_types.go

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package compute
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-openstack/api/v1alpha4"
+)
+
+// InstanceSpec defines the fields which can be set on a new OpenStack instance.
+//
+// InstanceSpec does not contain all of the fields of infrav1.Instance, as not
+// all of them can be set on a new instance.
+type InstanceSpec struct {
+	Name           string
+	Image          string
+	Flavor         string
+	SSHKeyName     string
+	UserData       string
+	Metadata       map[string]string
+	ConfigDrive    bool
+	FailureDomain  string
+	RootVolume     *infrav1.RootVolume
+	Subnet         string
+	ServerGroupID  string
+	Trunk          bool
+	Tags           []string
+	SecurityGroups []string
+	Networks       []infrav1.Network
+}
+
+// InstanceIdentifier describes an instance which has not necessarily been fetched.
+type InstanceIdentifier struct {
+	ID   string
+	Name string
+}
+
+// InstanceStatus represents instance data which has been returned by OpenStack.
+type InstanceStatus struct {
+	server *servers.Server
+}
+
+func NewInstanceStatusFromServer(server *servers.Server) *InstanceStatus {
+	return &InstanceStatus{server}
+}
+
+type networkInterface struct {
+	Address string  `json:"addr"`
+	Version float64 `json:"version"`
+	Type    string  `json:"OS-EXT-IPS:type"`
+}
+
+// InstanceNetworkStatus represents the network status of an OpenStack instance
+// as used by CAPO. Therefore it may use more context than just data which was
+// returned by OpenStack.
+type InstanceNetworkStatus struct {
+	addresses map[string][]networkInterface
+}
+
+func (is *InstanceStatus) ID() string {
+	return is.server.ID
+}
+
+func (is *InstanceStatus) Name() string {
+	return is.server.Name
+}
+
+func (is *InstanceStatus) State() infrav1.InstanceState {
+	return infrav1.InstanceState(is.server.Status)
+}
+
+func (is *InstanceStatus) SSHKeyName() string {
+	return is.server.KeyName
+}
+
+// APIInstance returns an infrav1.Instance object for use by the API.
+func (is *InstanceStatus) APIInstance() (*infrav1.Instance, error) {
```

Comment:
looks like we have more param for Metadata, ConfigDrive etc for Infrav1.Instance
will it be inited here or somewhere else?

Reply:
That's the intention, although I don't currently think we'll add any paramaters to APIInstance() specifically. We will almost definitely add parameters to NetworkStatus(), though.